### PR TITLE
Enhancement: End date validation

### PR DESCRIPTION
**Pull request by kayladnls**
This PR: 
- [x] adds validation that End Date must be after Start Date. 
- [x] requires phpunit as a dev dependency
- [x] specifies 5.6 as a dependency
- [x] covers EventMapper with tests

Jira Ticket: 
https://joindin.jira.com/browse/JOINDIN-676

### CODE REVIEW

**Comment by lornajane** [comment]
ok to test

**Comment by kayladnls** [comment]
Not sure why jenkins is failing. I'm not familiar with is as a CI solution. It's showing a checkstyle violation but it says "0 Violations, 0 Violating files". 

image

**Comment by dstockto** [comment]
@kayladnls It appears that phpunit died during running unit tests.

**Comment by magicmonkey** [comment]
Jenkins, test this please!

(I'm debugging the Jenkins build step for this, please ignore)

**Comment by magicmonkey** [comment]
Jenkins, test this please?

**Comment by akrabat** [comment]
The phpunit on the build server is too old for `expectException()`, which I think came in with PHPUnit 5.2.

**Comment by akrabat** [comment]
I have PHPUNit 5.1.4 here: https://i.19ft.com/7602fc00.png

**Comment by magicmonkey** [comment]
We get this fatal error in the PHP logs:

```
[21-Feb-2016 11:14:55 UTC] PHP Fatal error:  Call to undefined method EventMapperTest::expectException() in /var/lib/jenkins/jobs/api-pr-build/workspace/tests/src/models/EventMapperTest.php on line 7
[21-Feb-2016 11:14:55 UTC] PHP Stack trace:
[21-Feb-2016 11:14:55 UTC] PHP   1. {main}() /usr/bin/phpunit:0
[21-Feb-2016 11:14:55 UTC] PHP   2. PHPUnit_TextUI_Command::main() /usr/bin/phpunit:513
[21-Feb-2016 11:14:55 UTC] PHP   3. PHPUnit_TextUI_Command->run() phar:///usr/bin/phpunit/phpunit/TextUI/Command.php:105
[21-Feb-2016 11:14:55 UTC] PHP   4. PHPUnit_TextUI_TestRunner->doRun() phar:///usr/bin/phpunit/phpunit/TextUI/Command.php:154
[21-Feb-2016 11:14:55 UTC] PHP   5. PHPUnit_Framework_TestSuite->run() phar:///usr/bin/phpunit/phpunit/TextUI/TestRunner.php:425
[21-Feb-2016 11:14:55 UTC] PHP   6. PHPUnit_Framework_TestSuite->run() phar:///usr/bin/phpunit/phpunit/Framework/TestSuite.php:747
[21-Feb-2016 11:14:55 UTC] PHP   7. PHPUnit_Framework_TestCase->run() phar:///usr/bin/phpunit/phpunit/Framework/TestSuite.php:747
[21-Feb-2016 11:14:55 UTC] PHP   8. PHPUnit_Framework_TestResult->run() phar:///usr/bin/phpunit/phpunit/Framework/TestCase.php:726
[21-Feb-2016 11:14:55 UTC] PHP   9. PHPUnit_Framework_TestCase->runBare() phar:///usr/bin/phpunit/phpunit/Framework/TestResult.php:616
[21-Feb-2016 11:14:55 UTC] PHP  10. PHPUnit_Framework_TestCase->runTest() phar:///usr/bin/phpunit/phpunit/Framework/TestCase.php:770
[21-Feb-2016 11:14:55 UTC] PHP  11. ReflectionMethod->invokeArgs() phar:///usr/bin/phpunit/phpunit/Framework/TestCase.php:907
[21-Feb-2016 11:14:55 UTC] PHP  12. EventMapperTest->testCannotCreateIfEndDateIsBeforeStartDate() phar:///usr/bin/phpunit/phpunit/Framework/TestCase.php:907
```

**Comment by akrabat** [comment]
i.e. although phpunit is now specified as a composer dependency and hence it's in vendor/bin, the phing task in build.xml will need to be updated.

@magicmonkey can you update the phpunit on the server to the latest in the 5.2 series please?

**Comment by magicmonkey** [comment]
Yep, done:
```
kevin@joindin:~$ phpunit --version
PHPUnit 5.2.9 by Sebastian Bergmann and contributors.
```
although I've not done it for the dev VM (and I'm not going to).

Jenkins: feel free to test this again.

**Comment by magicmonkey** [comment]
Jenkins, test this please

**Comment by akrabat** [comment]
If we're going to put phpunit in the composer require-dev section, then I suspect that we'll need to update the launch script to remove it and its dependencies on deploy. We should also do the same for web2 and legacy if we go down this route.

Thoughts?

**Comment by magicmonkey** [comment]
Seems to have worked.

**Comment by magicmonkey** [comment]
Yeah, it needs a bit of thinking about as it's a change to how we do things and there's a few moving parts in there (composer, Jenkins, phpunit, deploy script, dev VM).  Probably beyond the scope of this PR, and worth keeping this PR only as the functional changes as opposed to anything systemic in the interest of only having small things change at once?

File: src/models/EventMapper.php [hunk 1]
```diff
@@ -763,7 +763,7 @@ public function createEvent($event, $auto_approve = false)
     {
 
         // Sanity check: ensure all mandatory fields are present.
-        $mandatory_fields          = array(
+        $mandatory_fields          = [
```
Comment (akrabat) [inline src/models/EventMapper.php]: No need for this change.

File: src/models/EventMapper.php [hunk 2]
```diff
@@ -771,12 +771,17 @@ public function createEvent($event, $auto_approve = false)
...
             'contact_name',
-        );
+        ];
+
         $contains_mandatory_fields = !array_diff($mandatory_fields, array_keys($event));
         if (!$contains_mandatory_fields) {
             throw new Exception("Missing mandatory fields");
         }
 
+        if (strtotime($event['start_date']) > strtotime($event['end_date'])) {
```
Comment (akrabat) [inline src/models/EventMapper.php]: This code is in the wrong place. All other validation checks are in the `EventsController` here: https://github.com/joindin/joindin-api/blob/master/src/controllers/EventsController.php#L202-L206

Also, note that the `start_date` and `end_date` are already unix timestamps, so don't need to be strtotime'd

File: tests/bootstrap.php [hunk 3]
```diff
@@ -1,3 +1,4 @@
 <?php
 // Load and register autoloader
 require_once __DIR__ . '/../src/inc/Autoloader.php';
+require_once __DIR__ . '/../vendor/autoload.php';
```
Comment (akrabat) [inline tests/bootstrap.php]: This line isn't needed, I think?

File: tests/src/phpunit.xml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        <directory>./</directory>
+    </testsuite>
+    <logging>
+        <log type="coverage-html" target="../build/coverage" title="joindin"
+             charset="UTF-8" yui="true" highlight="true"
+             lowUpperBound="35" highLowerBound="70"/>
+        <log type="coverage-clover" target="../build/logs/clover.xml"/>
+        <log type="junit" target="../build/logs/junit.xml" logIncompleteSkipped="false"/>
+    </logging>
+</phpunit>
```
Comment (akrabat) [inline tests/src/phpunit.xml]: I don't think this file is needed, given that we have `tests/phpunit.xml`?

**Comment by akrabat** [comment]
Hi @kayladnls 

Thanks for the PR. I've added some comments to the code and also as you can see above, we'd prefer not to deal with phpunit dependency changes as part of this PR.  Looking at it, I think that removing commits `8c7aa16`, `eda9e84`, and `7dc2561` will take out the PHPUnit installation.

File: src/models/EventMapper.php (hunk 2 above)
Comment (kayladnls) [inline src/models/EventMapper.php]: I see. I can make that change.

**Comment by kayladnls** [comment]
@akrabat Roger. I'll pull out the composer require stuff and go from there. Thanks!

**Comment by lornajane** [comment]
What's the status of this PR?  @kayladnls are you still working on it?

**Comment by lornajane** [comment]
Closing as I don't think we're in a position to accept yet and I think this PR is inactive, but I'd still really like this feature in the project at some point!

**Pull request closed without merging**